api/proto/config/storage.proto

```diff
@@ -106,3 +106,48 @@ message Drive {
   // A value of 0 will indicate that no resizing is required
   int64 maxsizebytes = 10;
 }
+
+
+message ContentTree {
+  UUIDandVersion uuidandversion = 1;
+  string dsId  = 2; // effectively pointer/key into dsConfigs
+
+  string URL = 3;
+  Format iformat = 4;
+
+  // the following is only used for individual blobs, if this message references a group of blobs
+  // e.g. in the case of OCI -- this information is expected to be provided by a top level blob
+  // that this message points to via its URL
+  string sha256 = 5;
+  uint64 sizeBytes = 6; // used for capping resource consumption eve for OCI & BLOB_TREE
+  SignatureInfo siginfo = 7;
+}
+
+enum VolumeAccessProtocols {
+  VAP_NONE = 0;
+  VAP_9P = 1;
+  VAP_NFS = 2;
+  VAP_SCSI = 3;
+  VAP_VIRTIO = 4;
+}
+
+message Volume {
+  UUIDandVersion uuidandversion = 1;
+
+  // root of the blob tree required to construct this Volume (can be null)
+  string contentStoreId = 2;
```

Comment:
I think we want a more flexible way to describe different origins as specified in the document.
Thus having a separate ContentOrigin message embedded here which can be used to describe different ones without having to add new attributes in the Volume message itself.

Reply:
I read the document again, I believe the proto definitions in storage.proto are faithfully derived from the document. contentorigin can be derived as follows: if contentTreeID is null, then contentOriginType is "blank", else it is "download". The contentTree (corresponding to contentTreeID) will specify the contentDownloadOrigin(dsId, sha etc.). If that does not suffice, can you please update the 'config' section of the document. I can then change the proto definitions to match those.

Reply:
I believe that is at least awkward if not insufficient as we later need to be able to introduce "protocol translators" to some external protocol for storage. The origin stuff was added to be able to support that without having to add a bunch of additional attributes in the volume itself.

Reply:
@nordmark I am ok with separating out contentorigin. Please update the 'config' portion of the document (https://wiki.lfedge.org/display/EVE/Volume+%28and+Content%29+device+API)  so I can match proto definitions with what is in the document.

Reply:
I updated the document. I don't know how easy it is to tell the diffs - I got rid of the downloadorigin message on the info side. Might need to add a contentTree on the info side but can't tell until I can more easily look at the whole API

Reply:
@nordmark just saw the update.

message volumeContentOrigin {
        volumeContentOriginType type = 1;
        VolumeType voltype // describes the type of the constructed volume (note that "EMPTY" is not used; that is the "BLANK" type)
        string downloadContentStoreID = 2; // where we get DOWNLOAD types from
        // TBD More optional fields for other originTypes
}

should 'downloadContentStoreID' be 'downloadContentTreeID'? My understanding was volume refers to contentTree, which intern refers to datastore.

The following is currently already defined for contentTree on the info side.

message ZInfoContentTree {
  string uuid = 1;
  string displayName = 2;
  ContentResources resources = 3;
  UsageInfo usage = 4;  // Same info as for volumes
  repeated string componentShaList = 5;
}

Reply:
> should 'downloadContentStoreID' be 'downloadContentTreeID'? My understanding was volume refers to contentTree, which intern refers to datastore.

Yes.